rs/execution_environment/src/scheduler.rs

```diff
@@ -2461,6 +2461,17 @@ fn is_next_method_chosen(
     may_schedule_heartbeat: bool,
     may_schedule_global_timer: bool,
 ) -> bool {
+    // If the task on the front of the task queue is hook,
+    // it must be executed next.
+    if canister
+        .system_state
+        .task_queue
+        .front()
+        .map_or(false, |task| task.is_hook())
+    {
+        return true;
+    }
```

Comment:
When the canister is executed we always check if we should execute the task by calling `is_next_method_chosen` and only if there is no method on the task queue after it, a message will be executed. This way we enforce that the hook is always executed prior to messages as well.